readme.md
EZDSL 3.10
==========

_Easy classical data structures for Delphi_

Introduction
------------

The EZDSL units provide an OOP interface for classical data structures
for Delphi: stacks, queues, priority queues, lists, binary trees, hash
tables, and so forth.

**DANGER, WILL ROBINSON!**

These units were written originally for Delphi 1. You remember that: 16-bit,
shortstrings, still very pointer-bound. Over the years they've been upgraded to
32-bit with later Delphis, but they still maintain their roots. If I were
writing Delphi these days (I don't), I'd chuck all the pointer-based stuff 
and make them fully OOP, Unicode string capable, generic, with some patterns 
thrown in for good measure. But they are what they are: if you've been using 
them over the years, here they are for XE+ Delphis for code written "old-style".

So:

- if you are using long strings, you'll have to take care of the problems 
casting long strings to pointers and back again (the reference counting needs 
to be taken care of by code you write);
- there is no 64-bit support since several units use assembly (either 16- or 32-bit);
- there's issues with the red-black binary tree (as in, it's still a binary tree, 
but the balancing algorithms are subtly broken, it needs rewriting);
- pretty much all source files use obtuse, acronymic, 8.3 naming because of the 
continued support for Delphi 1.


Documentation
-------------

See EZDSL3.MD. 

Licensing
---------

Copyright (c) 1993-2015, Julian M Bucknall
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are 
met:

1. Redistributions of source code must retain the above copyright 
notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright 
notice, this list of conditions and the following disclaimer in the 
documentation and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its 
contributors may be used to endorse or promote products derived from 
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT 
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, 
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY 
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



CHANGES
EZDSL Change Log
======================================================================

     Ver
Num  Fixed File          Typ  Description
----------------------------------------------------------------------
001  1.00  EZDSLLST.PAS  fix  Iterate doesn't return nil as default
           EZDSLDBL.PAS
           EZDSLSKP.PAS
002  1.00  EZDSLLST.PAS  fix  @Self is used instead of Self
           EZDSLDBL.PAS
           EZDSLSKP.PAS
           EZDSLBTR.PAS
003  2.00  EZDSLCOL.PAS  enh  Raise exception on bad case switch
           EZDSLBTR.PAS
004  2.00  various       enh  Removal of else clauses for certain ifs
005  2.00  EZDSLCOL.PAS  fix  TEZCollection.Clone always duped data
006  2.00  EZDSLBSE.PAS  enh  DupData can be set at all times
007  2.00  EZDSLLST.PAS  fix  Sorted lists are always produced from
           EZDSLDBL.PAS       sorted lists
008  2.00  EZDSLCOL.PAS  fix  TEZCollection.Assign always duped data
009  3.00  EZDSLHSH.PAS  fix  The hash method can sometimes return a
                              negative number
010  3.00  EZDSLBAR.PAS  enh  Added a boolean array class
011  3.01  EZDSLLST.PAS  fix  If Search successfully finds a data
           EZDSLDBL.PAS       object in a single & double linked list
                              it would not return the cursor.
012  3.01  EZDSLOPT.INC  fix  Support for Delphi 4 & 5
           EZDSLBTR.PAS
           EZDSLTHD.PAS
013  3.02  EZDSLBAR.PAS  fix  Numerous fixes to boolean array class
014  3.02  EZDSLOPT.INC  fix  Support for Delphi 6


======================================================================

001  1.00  EZDSLLST.PAS  fix  Iterate doesn't return nil as default
           EZDSLDBL.PAS
           EZDSLSKP.PAS
----------------------------------------------------------------------
The Iterate method is supposed to return nil if none of the calls to
the Action routine return false. The fix is the same to all of the
Iterate routines: add the following line as the last statement of
the routine:

       end;
     Result := nil;                                              {!!.01}
   end;
 {--------}


002  1.00  EZDSLLST.PAS  fix  @Self is used instead of Self
           EZDSLDBL.PAS
           EZDSLSKP.PAS
           EZDSLBTR.PAS
----------------------------------------------------------------------
In the various Iterate methods defined in the above source files the
container object is passed to the Action function as @Self. It should
be Self. Whereever @Self is found in the above files, replace with
Self.


003  2.00  EZDSLCOL.PAS  enh  Raise exception on bad case switch
           EZDSLBTR.PAS
----------------------------------------------------------------------
As an enhancement, any case statement without an else clause was
expanded to include one. The else clause would raise an exception to
indicate that an invalid switch value was given:

  case ... of
    ...
  else                                                   {!!.01}
    RaiseError(escBadCaseSwitch);                        {!!.01}
  end;{case}


004  2.00  various       enh  Removal of else clauses for certain ifs
----------------------------------------------------------------------
Certain if statements were constructed thus:

  if there is an error condition then
    raise an exception
  else
    do something

These if..then..else statements were rewritten as such:

  if there is an error condition then
    raise an exception
  do something

since is an exception is raised no other statements will be executed
after the statement that raises the exception. These changes have in
general NOT been marked.


005  2.00  EZDSLCOL.PAS  fix  TEZCollection.Clone always duped data
----------------------------------------------------------------------
The TEZCollection.Clone method always duplicated the data, whether the
collection was a data owner or not. The fix is to test for data
ownership, if true the data is duplicated, if false the pointer value
is copied over:

constructor TEZCollection.Clone(...);
  ...
  begin
    ...
    if not OldColl.IsEmpty then
      for i := 0 to pred(OldColl.Count) do
        begin
          if DataOwner then                                    {!!.01}
            NewData := DupData(OldColl.Items[i])               {!!.01}
          else                                                 {!!.01}
            NewData := OldColl.Items[i];                       {!!.01}
          ...


006  2.00  EZDSLBSE.PAS  enh  DupData can be set at all times
----------------------------------------------------------------------
The restriction that DupData can only be set when the container is
empty has been removed. The change is to TAbstractContainer.SetDupData
in the EZDSLBSE.PAS unit.

procedure TAbstractContainer.SetDupData(NewFunc : TDupDataFunc); {!!.01}
  begin
    if Assigned(NewFunc) then
         FDupData := NewFunc
    else FDupData := EZAbstractDupData;
  end;


007  2.00  EZDSLLST.PAS  fix  Sorted lists are always cloned from
           EZDSLDBL.PAS       sorted lists
----------------------------------------------------------------------
If a linked list is sorted then only sorted lists can be cloned from
it and vice-versa. The linked lists' Clone constructors have been
enhanced so that the cloned list is sorted if the NewCompare parameter
is not nil. The change is to the Clone constructor of both TLinkList
and TDList.

TLinkList.Clone:

    if Assigned(NewCompare) then                               {!!.01}
      Compare := NewCompare                                    {!!.01}
    else                                                       {!!.01}
      Compare := OldList.Compare;                              {!!.01}
    ...
    if Assigned(NewCompare) then                               {!!.01}
         OldList.Iterate(SListSortedClone, false, Self)
    else OldList.Iterate(SListClone, false, Self);

TDList.Clone:

    if Assigned(NewCompare) then                               {!!.01}

EZDSLCts.PAS
{===EZDSLCts==========================================================

Constants for the EZDSL (Delphi Structures Library) unit.

EZDSLCTS is Copyright (c) 1993-2011 by  Julian M. Bucknall

VERSION HISTORY
02Jul11 JMB 3.04 Release for Delphi XE
12Feb02 JMB 3.03 Release for Delphi 6
24Oct99 JMB 3.02 Release for Delphi 4 & 5
14May98 JMB 3.01 Inclusion of TBooleanArray string codes
19Apr98 JMB 3.00 Major new version, release for Delphi 3
24May96 JMB 2.01 extra string codes
13Mar96 JMB 2.00 release for Delphi 2.0
18Jun95 JMB 1.00 initial release
=====================================================================}
{ Copyright (c) 1993-2002, Julian M. Bucknall. All Rights Reserved   }

unit EzdslCts;

interface

{Note: the actual strings are found in EZDSLCTS.RC and
       EZDSLCTS.R16/.R32}

const
  {Start offset for string constants}
  {Note: this can be altered to suit the application}
  ezdsStartOffset = $62D5;

  {Error string constants}
  escTooManyItems   = ezdsStartOffset+ 1;
  escInsInvalidHere = ezdsStartOffset+ 2;
  escDelInvalidHere = ezdsStartOffset+ 3;
  escInsertDup      = ezdsStartOffset+ 4;
  escCannotMoveHere = ezdsStartOffset+ 5;
  escNoCompare      = ezdsStartOffset+ 6;
  escNoDupData      = ezdsStartOffset+ 7;
  escNoDisposeData  = ezdsStartOffset+ 8;
  escBadSource      = ezdsStartOffset+ 9;
  escIndexError     = ezdsStartOffset+10;
  escBadCaseSwitch  = ezdsStartOffset+11;
  escKeyNotFound    = ezdsStartOffset+12;
  escTableFull      = ezdsStartOffset+13;
  escSortNeedsCmp   = ezdsStartOffset+14;
  escCmpNeeded      = ezdsStartOffset+15;
  escBadBooleanInx  = ezdsStartOffset+16;


  {Assertion string constants}
  ascFreeNilNode    = ezdsStartOffset+50;
  ascNewNodeSize0   = ezdsStartOffset+51;
  ascFreeNodeSize0  = ezdsStartOffset+52;
  ascEmptyExamine   = ezdsStartOffset+53;
  ascEmptyPop       = ezdsStartOffset+54;
  ascDeleteEdges    = ezdsStartOffset+55;
  ascExamineEdges   = ezdsStartOffset+56;
  ascInsertEdges    = ezdsStartOffset+57;
  ascReplaceEdges   = ezdsStartOffset+58;
  ascAlreadyAtEnd   = ezdsStartOffset+59;
  ascAlreadyAtStart = ezdsStartOffset+60;
  ascCannotJoinHere = ezdsStartOffset+61;
  ascCannotJoinData = ezdsStartOffset+62;
  ascSplitEdges     = ezdsStartOffset+63;
  ascOutOfRange     = ezdsStartOffset+64;
  ascExamineLeaf    = ezdsStartOffset+65;
  ascBadSkipLevel   = ezdsStartOffset+66;
  ascIsSortedList   = ezdsStartOffset+67;
  ascIsNotSortedList= ezdsStartOffset+68;
  ascBadCapacity    = ezdsStartOffset+69;
  ascNilArray       = ezdsStartOffset+70;
  ascNotSameSize    = ezdsStartOffset+71;

implementation

end.
